runtime/major_gc.c

```diff
@@ -1402,6 +1402,8 @@ void caml_opportunistic_major_collection_slice(intnat howmuch)
 
 void caml_major_collection_slice(intnat howmuch)
 {
+  caml_domain_state *d = Caml_state;
+  uintnat major_slice_epoch = atomic_load (&caml_major_slice_epoch);
```

Comment:
Is there a reason for the SC ordering on this variable here and elsewhere and not a weaker one? (I am asking from the point of view that using the weakest ordering makes the code easier to understand since relaxed and acq/rel ordering are often self-documenting; when I see the use of SC ordering I need some lengthy comment explaining what is being synchronised and how.)

Reply:
For code that is not performance sensitive, my approach had been to pick the ordering that doesn't require complicated reasoning. Hence the choice of SC atomics there. When I see acquire/release atomics my takeaway would be that "someone has tried to optimise this and so this must be performance sensitive". What approach do we want to take in the runtime code. @xavierleroy?

FWIW release/acquire ordering should be sufficient here.

Reply:
I'm in agreement with @kayceesrk here, when in doubt (and it isn't somewhere very hot) then SC is fine. We should err to be too strong than too weak.

Reply:
I think the issue is that it is not obvious to me what ordering you need to reason that the code is correct. (For instance, is it innocuous to reload `caml_minor_collections_count` inside `advance_global_major_slice_epoch`, and why?)

I do not think one should "err" one way or another, I think it is better if the reason why it is correct is laid out for the reader without them having to guess. I simply point out that in the event where you only need acq-rel or relaxed ordering, it is often the case that using the most precise ordering makes the code (more) self-documenting. (If looking for a general principle, reasoning about correctness takes precedence over optimizations; if some code has in addition been noticed to be performance-sensitive then I think it makes sense to write it in a comment.) Note that I do not know if we are in this situation here given the interaction with the `atomic_exchange`, an alternative could be to add comments.

Keep in mind that this is in a context where atomic variables accumulate in the multicore runtime, it has been hard to read in other places, and there has been synchronisation issues in the past for things that looked simple.

Reply:
> If looking for a general principle, reasoning about correctness takes precedence over optimizations

precisely why I chose SC here

Reply:
I see the merits of @gadmm's position, but I would still side with @kayceesrk on this point.

Reply:
For some reason I never got the e-mail notification of your reply @kayceesrk, sorry for the delay.

> precisely why I chose SC here

Earlier you said that the principle applied here was an acquire-release pattern, now you say that using SC ordering helps with the reasoning. Which one of the two is correct? (From how I learnt to reason about parallel code, there is a contradiction here---but perhaps I am just missing something about what you mean which is obvious to you, that I still do not see.)

What I see is a function that reloads an atomic variable for no apparent reason, which is a yellow flag in parallel code to me. It is possible that the reloading is harmless, for simple reasons even, but I am not going to spend time to try to convince myself that it is so, because when I asked to clarify the code somehow, you declined.

In any case this was only meant to be a small follow-up to the review I previously did on the draft of this PR, I did not want to make an example of the PR about this code style issue (which reminded me of frustrations when trying to review the multicore runtime) nor block the PR on it.

Reply:
> Earlier you said that the principle applied here was an acquire-release pattern,

I don't think I mentioned that I am using acquire/release reasoning here. I mention in https://github.com/ocaml/ocaml/pull/11750/files#r1046733634 that acquire/release is sufficient, but not that I am using acquire/release reasoning. I must be missing something. 

> In any case this was only meant to be a small follow-up to the review I previously did on the draft of this PR, I did not want to make an example of the PR about this code style issue (which reminded me of frustrations when trying to review the multicore runtime) nor block the PR on it.

Thanks.